File: Python/19_thermometer_1/ds18b20.py
#!/usr/bin/env python


def ds18b20Read():
	tfile = open("/sys/bus/w1/devices/28-0000062abd71/w1_slave")
	text = tfile.read()
	tfile.close()
	secondline = text.split("\n")[1]
	temperaturedata = secondline.split(" ")[9]
	temperature = float(temperaturedata[2:])
	temperature = temperature / 1000
	temperature = round(temperature, 2)
	print temperature
	return temperature

File: C/19_thermometer_1/ds18b20.c
#include "ds18b20.h"

extern float ds18b20Read(void)
{
	float temp;
	int i, j;
    int fd;
	int ret;

	char buf[BUFSIZE];
	char tempBuf[5];
	
	fd = open("/sys/bus/w1/devices/28-8000000178d0/w1_slave", O_RDONLY);

	if(-1 == fd){
		perror("open device file error\n");
		return -1;
	}

	while(1){
		ret = read(fd, buf, BUFSIZE);
		if(0 == ret){
			break;	
		}
		if(-1 == ret){
			if(errno == EINTR){
				continue;	
			}
			perror("read()\n");
			close(fd);
			return 1;
		}
	}

	for(i=0;i<sizeof(buf);i++){
		if(buf[i] == 't'){
			for(j=0;j<sizeof(tempBuf);j++){
				tempBuf[j] = buf[i+2+j]; 	
			}
		}	
	}

	temp = (float)atoi(tempBuf) / 1000;

//	printf("%.3f C\n",temp);

	close(fd);

	return temp;
}
